src/FilenameParsing/AbstractFileIDHelper.php

```diff
@@ -0,0 +1,153 @@
+    {
+        return $this->swapExtension($filename, $variant, self::EXT_VARIANT);
+    }
+
+    /**
+     * Get the variant filename with the extension rewritten to be the same as the original file extension.
+     *
+     * @param string $filename Original filename without variant
+     */
+    public function restoreOriginalExtension(string $filename, string $variant): string
+    {
+        return $this->swapExtension($filename, $variant, self::EXT_ORIGINAL);
+    }
+
+    /**
+     * Get the original file's filename with the extension rewritten to be the same as either the original
+     * or the variant extension.
+     *
+     * @param string $filename Original filename without variant
+     * @param int $extIndex One of self::EXT_ORIGINAL or self::EXT_VARIANT
+     */
+    private function swapExtension(string $filename, string $variant, int $extIndex): string
+    {
+        // If there's no variant at all, we can rewrite the filenmane
+        if (empty($variant)) {
+            return $filename;
+        }
+
+        // Split variant string in variant list
+        $subVariants = explode('_', $variant);
+
+        // Split our filename into a filename and extension part
+        if (!preg_match('/(?<basename>.+)\.(?<ext>[a-z\d]+)$/i', $filename, $matches)) {
+            return $filename;
+        }
+        $filenameWithoutExtension = $matches['basename'];
+        $extension = $matches['ext'];
+
+        // Loop our variant list until we find our special file extension swap variant
+        // Reverse the list first so the variant extension we find is the last extension rewrite variant in a chain
+        $extSwapVariant = preg_quote(self::EXTENSION_REWRITE_VARIANT, '/');
+        foreach (array_reverse($subVariants) as $subVariant) {
+            if (preg_match("/^$extSwapVariant(?<base64>.+)$/", $subVariant, $matches)) {
+                // This array always contain 2 values: The original extension at index 0 and the variant extension at index 1
+                /** @var array $extensionData */
+                $extensionData = Convert::base64url_decode($matches['base64']);
+                $extension = $extensionData[$extIndex];
+                break;
+            }
+        }
+
+        return $filenameWithoutExtension . '.' . $extension;
+    }
+
+    public function cleanFilename($filename)
+    {
+        // Swap backslash for forward slash
+        $filename = str_replace('\\', '/', $filename ?? '');
+
+        // Since we use double underscore to delimit variants, eradicate them from filename
+        return preg_replace('/_{2,}/', '_', $filename ?? '');
+    }
+
+    public function lookForVariantRecursive(): bool
+    {
+        return false;
+    }
+
+    abstract protected function getFileIDBase($shortFilename, $fullFilename, $hash, $variant): string;
+
+    abstract protected function validateFileParts($filename, $hash, $variant): void;
```

Comment:
These accept more arguments than the current concrete classes use to make it easier to have additional subclasses with their own validation or fileIDBase implementations.